system/datatypes/mailinglist.php

```diff
@@ -1,406 +1,433 @@
+		$addressCollection = $this->xml->findXPath($xpath);
 		$address = $addressCollection->first();
 		if (!$address instanceof HyphaDomElement) {
 			$this->xml->unlock();
 			notify('error', __('invalid-code'));
-
 			return null;
 		}
 
-		if ($address->getAttribute('status') == self::ADDRESS_STATUS_CONFIRMED) {
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) == self::ADDRESS_STATUS_CONFIRMED) {
 			$this->xml->unlock();
 			notify('success', ucfirst(__('ml-successfully-subscribed')));
 			return ['redirect', $this->constructFullPath($this->pagename)];
 		}
 
-		$address->setAttribute('status', self::ADDRESS_STATUS_CONFIRMED);
-		$address->setAttribute('unsubscribe-code', $this->constructCode());
+		$address->setAttribute(self::FIELD_NAME_STATUS, self::ADDRESS_STATUS_CONFIRMED);
+		$address->setAttribute(self::FIELD_NAME_UNSUBSCRIBE_CODE, $this->constructCode());
 		/** @var HyphaDomElement $addresses */
-		$addresses = $this->getDoc()->getOrCreate('addresses');
+		$addresses = $this->xml->find(self::FIELD_NAME_ADDRESSES_CONTAINER);
 		$addresses->append($address);
 		$this->xml->saveAndUnlock();
 
 		notify('success', ucfirst(__('ml-successfully-subscribed')));
 		return ['redirect', $this->constructFullPath($this->pagename)];
 	}
 
-	private function unsubscribeAction() {
+	private function unsubscribeAction(HyphaRequest $request) {
 		$code = isset($_GET['code']) ? $_GET['code'] : null;
 		if (null == $code) {
 			notify('error', __('missing-arguments'));
-
 			return null;
 		}
 
 		$this->xml->lockAndReload();
+
 		/** @var NodeList $addressCollection */
-		$addressCollection = $this->getDoc()->findXPath('//address[@unsubscribe-code="' . $code . '"]');
+		$xpath = '//' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_UNSUBSCRIBE_CODE . '=' . xpath_encode($code) . ']';
+		$addressCollection = $this->xml->findXPath($xpath);
 		$address = $addressCollection->first();
 		if (!$address instanceof HyphaDomElement) {
 			$this->xml->unlock();
 			notify('error', __('invalid-code'));
-
 			return null;
 		}
 
-		if ($address->getAttribute('status') == self::ADDRESS_STATUS_PENDING) {
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) == self::ADDRESS_STATUS_PENDING) {
 			// weird case, should never happen, let's handle it anyway
 			$this->xml->unlock();
 			notify('error', __('ml-mail-has-invalid-status'));
-
 			return null;
 		}
 
-		if ($address->getAttribute('status') != self::ADDRESS_STATUS_UNSUBSCRIBED) {
-			$address->setAttribute('status', self::ADDRESS_STATUS_UNSUBSCRIBED);
-			$address->removeAttribute('email');
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) != self::ADDRESS_STATUS_UNSUBSCRIBED) {
+			$address->setAttribute(self::FIELD_NAME_STATUS, self::ADDRESS_STATUS_UNSUBSCRIBED);
+			$address->setAttribute(self::FIELD_NAME_UNSUBSCRIBE_EMAIL, $address->getAttribute(self::FIELD_NAME_EMAIL));
```

Comment:
What is the idea behind this extra field? If we want to keep the e-mail address, why not just in the regular e-mail field?

Reply:
If you keep it in the same field and the same email address would subscribe, that recor would be re-used.

Reply:
Is that problematic? And if so, you could also change the subscription code to not reuse an UNSUBSCRIBED status record?

Reply:
true, but  don't see how that would be an improvement with the current implementation.

Reply:
What if I were to move it to an entire other section? That would remove the extra field and prevent any problems.

Reply:
That also seems a bit convoluted. I would just keep it in the same space, to keep the structure simple and equal, and adapt other code to match. Most other code that deals with subscriptions already has to check for status anyway, so that is not particularly hard, right?

We could also consider preserving the current behaviour (removing email) in this PR, and move this change into a separate PR (a separate commit would be a good idea in any case, since it is quite a signficant behaviour change).

Reply:
Then we're adding more complexity without any added value. Why is the extra field a problem?

Reply:
Because, IMHO, it adds complexity (e.g. the display code must now handle two fields for e-mail rather than one).

Reply:
.... I've already retracted my previous comments. sorry. It just didn't click before.. Now it does. :-)

Reply:
fixed

Reply:
Yay! Also, deleting comments is sometimes annoying, since I still see them in e-mail notification but cannot use the link in the e-mail anymore to see the comments. Editing them to indicate they are no longr relevant might be better in such cases :-)